Persist current revision to a SQL-lite db and only update the stack if the revision and stack content changed

clangenb opened the pull request:
So our goal is to prevent unnecessary restarts as our restarts incur minimal offline time, which we absolutely want to prevent. Hence, this PR introduces the following optimizations:

* Only redeploy the stack iff the git revision **and** the particular stack's docker compose file have changed.
* Additionally, we persist both into an SQL light DB such that a restart of swarm cd won't cause a restart of all stacks.

clangenb committed:
Check git revision while updating the stack.

clangenb committed:
add sql_db for persisting git revisions

clangenb committed:
integrate persisting revision to db

clangenb committed:
add error handling in sql_db code

clangenb committed:
minor fixes

clangenb committed:
ui fixes

clangenb committed:
log the current revision in stack.go

clangenb committed:
add repos.yaml/stack.yaml to gitignore

clangenb committed:
fix readme regarding volume

clangenb committed:
delete docker compose

clangenb committed:
ignore docker compose

sanzoghenzo reviewed (commented):
This is great!

I've added my personal opinions on this: functionally it is OK (though I still have to test it), but I would reword some things.

clangenb committed:
Rename revision loading/saving

Co-authored-by: Andrea Ghensi <andrea.ghensi@gmail.com>

clangenb committed:
Rename revision loading/saving

Co-authored-by: Andrea Ghensi <andrea.ghensi@gmail.com>

clangenb committed:
Rename revision loading/saving

Co-authored-by: Andrea Ghensi <andrea.ghensi@gmail.com>

clangenb committed:
Rename revision loading/saving

Co-authored-by: Andrea Ghensi <andrea.ghensi@gmail.com>

clangenb committed:
rename sql_db to db

clangenb committed:
read db path from env var

clangenb commented on swarmcd/sql_db.go:
```diff
@@ -0,0 +1,72 @@
+package swarmcd
```
True, makes sense!

clangenb commented on swarmcd/sql_db.go:
```diff
@@ -0,0 +1,72 @@
+	_ "modernc.org/sqlite"
+)
+
+const dbFile = "/data/revisions.db"
```
Makes sense. It reads the path now from an env var, but defaults `/data/revision.db`.

clangenb committed:
add env var config to Readme

sanzoghenzo commented on README.md:
```diff
@@ -46,6 +46,15 @@ services:
+      - SWARMCD_DB=/data/revisions.db  # Pass DB file path as env variable
+  
+
+# Ensure that swarmdc_data persists across container restarts.
```
typo here, and also in the examples below

```suggestion
# Ensure that swarmcd_data persists across container restarts.
```

but... are all these comment in the compose file needed?
I believe they are pretty self explanatory for docker-compose/swarm users...

I think it would be better to add a specific section in the documentation like "data persistence" to explain the need for the volume and the optional settings via environment variable

clangenb replied:
They are not obvious to me, as I am not a pro in this domain XD. However, I guess you are right. 👍 I shall change it.

sanzoghenzo reviewed (commented):
Thank you for  the edits!
these are my last ramblings, I'll let you decide if they make sense 😉

clangenb committed:
Fix Typo

Co-authored-by: Andrea Ghensi <andrea.ghensi@gmail.com>

clangenb committed:
[Readme] remove unnecessary comments

clangenb committed:
[Readme] add minimal section regarding db config.

clangenb commented:
Thanks again for the detailed review. I have added the changes according to your feedback. 🚀

sanzoghenzo commented on README.md:
```diff
@@ -46,6 +46,13 @@ services:
       - ./repos.yaml:/app/repos.yaml:ro
       - ./stacks.yaml:/app/stacks.yaml:ro
+      - swarmcd_data:/data
+    environment:
+      - SWARMCD_DB=/data/revisions.db
```
```suggestion
```

sanzoghenzo commented on README.md:
```diff
@@ -57,6 +64,12 @@ docker stack deploy --compose-file docker-compose.yaml swarm-cd
+container restarts. By default, it stores data in data/revisions.db,
+but this can be changed via the `SWARMCD_DB` environment variable as 
+shown in the above docker-compose file.
+
```
Add a docker-compose sample with the changed environment variable here

sanzoghenzo commented on README.md:
```diff
@@ -100,10 +114,17 @@ services:
         target: /secrets/age.key
     environment:
       - SOPS_AGE_KEY_FILE=/secrets/age.key
+      - SWARMCD_DB=/data/revisions.db
```
```suggestion
```

sanzoghenzo commented on README.md:
```diff
@@ -214,9 +235,18 @@ services:
+      - swarmcd_data:/data
+    environment:
+      - SOPS_AGE_KEY_FILE=/secrets/age.key
+      - SWARMCD_DB=/data/revisions.db
```
```suggestion
```

sanzoghenzo reviewed (commented):
sorry for being so pedantic, but I ask you to do one last change to the docs to be as explicit as possible.

One thing I'm wondering just now, though: is it important to persist the database to disk?
Couldn't we just use an in-memory instance of the sqlite db?
I mean, the added complexity of the storage on disk will only save you a few re-deployment when you restart swarm-cd (which should be a zero operation since nothing changed), while the big benefit from this (as I see it) is to avoid the re-deployment every two minutes.

clangenb commented:
> sorry for being so pedantic, but I ask you to do one last change to the docs to be as explicit as possible.
> 
> One thing I'm wondering just now, though: is it important to persist the database to disk? Couldn't we just use an in-memory instance of the sqlite db? I mean, the added complexity of the storage on disk will only save you a few re-deployment when you restart swarm-cd (which should be a zero operation since nothing changed), while the big benefit from this (as I see it) is to avoid the re-deployment every two minutes.

No worries, being pedantic creates better code. In our case, we explicitly want to avoid redeployments due to restarts of swarm, as it will cause an unnecessary restart of the deployed service, which does incur minimal downtime - sometimes at unpredictable times.

However, maybe we could maybe disable that feature by default, and setting the env var will enable it. What do you think?

sanzoghenzo commented:
> as it will cause an unnecessary restart of the deployed service

Oh, now I'm the non-pro in this domain :sweat_smile: I thoight that re-issuing a docker stack deploy with the same exact file wouldn't result in a redeploy...

I see that there's the `--resolve-image` flag that defaults to `always`, and can be set to `changed` to only recreate the container if the image hash changes (be it a new tag or an update of the same tag)... maybe it can be enough to avoid or at least reduce the restarts?

This is not to turn down your contribution, it is just to analyze the situation and make it as efficient as possible (event if this is a small, simple addition, no code is the best code :rofl: )

clangenb commented:
hahaha, no worries. I am equally interested in finding the best aka most slim solution here. Let's have @mosonyi chime in here. He is the pro, and if he agrees that your suggestion might fix our issue, then I happy.

clangenb committed:
[swarmcd] add `stack_test.go` and first unit test

clangenb committed:
[swarmcd] also persist the hash of the stack content

clangenb committed:
[swarmcd] add unit tests for persisting revision and hash in to an in memory sqllite db

clangenb committed:
[swarmcd] create global db

clangenb committed:
[swarmcd] add global to track if db has been initialized

clangenb committed:
[swarmcd] close in memory db ad the end of the tests

clangenb committed:
[swarmcd] compare hashes of deployed filed

clangenb committed:
[gitignore] add config.yaml to identity

clangenb committed:
[gitignore] don't use global db references

clangenb committed:
[stack] add unit tests for rotateObjects

clangenb committed:
[stack] rotating configs works now with external: true instead of file.

clangenb committed:
[stack] skip writing to db with the new revision

clangenb committed:
[stack] hash the correct bytes to prevent a mismatch of hashes

clangenb committed:
[stack] better logs

clangenb committed:
[stack] better logs

clangenb committed:
[stack] better test setup

clangenb committed:
[stack] fix readme review comments

clangenb commented:
> > as it will cause an unnecessary restart of the deployed service
> 
> Oh, now I'm the non-pro in this domain 😅 I thoight that re-issuing a docker stack deploy with the same exact file wouldn't result in a redeploy...
> 
> I see that there's the `--resolve-image` flag that defaults to `always`, and can be set to `changed` to only recreate the container if the image hash changes (be it a new tag or an update of the same tag)... maybe it can be enough to avoid or at least reduce the restarts?
> 
> This is not to turn down your contribution, it is just to analyze the situation and make it as efficient as possible (event if this is a small, simple addition, no code is the best code 🤣 )

I finally had some time to lock at it again. :D

The `--resolve-image` flag would prevent redeployment of the stack if the image hash remains the same, but the stack config changes as far as I understand. I guess this would be an unwanted feature.

Additionally, we manage a lot of stacks in one repository, hence we had to realize that only checking the git revision is not enough. We need to check the hash of the contents of the stack's docker-compose file to identify if we want to redeploy a stack. I have added this extra feature now (including some general unit tests). I would be happy if you could take a look at the code again.

clangenb committed:
[database] use git hash as revision to prevent confusion

m-adawi commented:
Great work @clangenb thanks for this important feature

clangenb committed:
[stack] fix log about remaining at current revision

clangenb committed:
[stack] include all the read data into the hash

clangenb committed:
[stack] fix hash display for empty db

clangenb committed:
fix tests

clangenb committed:
be more explicit with the empty hash

sanzoghenzo reviewed (changes requested):
Hi, thanks again for the tremendous work!

I've merged some PRs, so there's some conflicts to resolve, sorry about that :sweat_smile: 

Apart from the comments in this review, I'm not convinced by the `rotateConfigsAndSecrets` and `rotateObjects` edits: they already calculate the hashes of configs and secrets, adding them to the names and editing the `composeMap` in-place; so you could just calculate the hash of `composeFileBytes` (once you extract it from the `writeStack` function).
I might be missing something, but I believe it would be enough to solve the issue at hand

clangenb committed:
Update swarmcd/database.go

Co-authored-by: Andrea Ghensi <andrea.ghensi@gmail.com>

clangenb committed:
Update swarmcd/stack.go

Co-authored-by: Andrea Ghensi <andrea.ghensi@gmail.com>

clangenb committed:
Update swarmcd/stack.go

Co-authored-by: Andrea Ghensi <andrea.ghensi@gmail.com>

clangenb committed:
Merge branch 'main' into cl/check-if-pulled-revision-matches-current

# Conflicts:
#	Dockerfile
#	swarmcd/stack.go
#	swarmcd/stack_test.go
#	swarmcd/swarmcd.go

clangenb commented:
> Hi, thanks again for the tremendous work!
> 
> I've merged some PRs, so there's some conflicts to resolve, sorry about that 😅
> 
> Apart from the comments in this review, I'm not convinced by the `rotateConfigsAndSecrets` and `rotateObjects` edits: they already calculate the hashes of configs and secrets, adding them to the names and editing the `composeMap` in-place; so you could just calculate the hash of `composeFileBytes` (once you extract it from the `writeStack` function). I might be missing something, but I believe it would be enough to solve the issue at hand

Yes, I am not so happy with my solution either. I have first checked the hashes at a different position, before I realized that I have to include the data read from the config too later. I think you are right, I will test your approach. Thanks for the hint!

clangenb committed:
fix merge error

clangenb committed:
better approach to calculate the stack hash

clangenb committed:
earlier return

clangenb committed:
fix unnecessary diff

clangenb committed:
better organize test file

clangenb committed:
fix merge errors

clangenb committed:
better naming

clangenb committed:
add line break in gitignore

clangenb committed:
extract should deploy method

clangenb committed:
return last revision if deployment was skipped

clangenb committed:
uniform log format

clangenb committed:
Merge branch 'main' into cl/check-if-pulled-revision-matches-current

# Conflicts:
#	web/routes.go

clangenb commented:
Thanks a lot for your review before. Indeed, your approach was much simpler. I think I have addressed all of your review comments, and I would be ready for another round. :D

clangenb committed:
remove that are no longer necessary after merging master

clangenb committed:
return error instead of nothing after opening database

sanzoghenzo commented on swarmcd/stack.go:
```diff
@@ -78,17 +98,31 @@ func (swarmStack *swarmStack) updateStack() (revision string, err error) {
+	}
+
+	if !swarmStack.shouldDeploy(writtenBytes, deployedStackHash, revision, lastRevision) {
+		return lastRevision, nil
```
should we update the db to the current revision anyway?

That way the next check would be quicker and exit at line 65

clangenb replied:
I absolutely see your point. However, we were thinking about this, and we were arguing whether it is better to see in the UI, which revision contained the last change, and don't display just the most recent one.

But I am not sure, which option is better here.

clangenb replied:
I circumvented this question by introducing an enriched `stackMetadata` type, which stores the `repoRevision` and the `deployedStack` revision.

clangenb committed:
Update swarmcd/stack.go

Co-authored-by: Andrea Ghensi <andrea.ghensi@gmail.com>

clangenb committed:
Update swarmcd/stack.go

Co-authored-by: Andrea Ghensi <andrea.ghensi@gmail.com>

clangenb commented on swarmcd/database.go:
```diff
@@ -0,0 +1,73 @@
+}
+
+// Load a stack's revision and hash
+func loadLastDeployedRevision(db *sql.DB, stackName string) (revision string, hash string, err error) {
```
Sure, can do that. 👍

clangenb committed:
introduce a struct for nicer interface

clangenb committed:
add fmtHash method to struct

clangenb committed:
fix sprintf statement

clangenb committed:
fix null-pointer upon first startup

clangenb committed:
persist current revision to DB even if we don't deploy

clangenb committed:
save deployment date

clangenb committed:
fix null pointer error

clangenb committed:
some renaming

clangenb committed:
more renaming

clangenb committed:
introduce stackDb abstraction

clangenb committed:
some renaming

clangenb commented:
Thanks to your pedantic feedback the code has been improved much more. It should be ready for another round of review.

clangenb commented on swarmcd/swarmcd.go:
```diff
@@ -30,17 +30,19 @@ func updateStackThread(swarmStack *swarmStack, waitGroup *sync.WaitGroup) {
 	stackStatus[swarmStack.name].Error = ""
-	stackStatus[swarmStack.name].Revision = revision
-	logger.Info(fmt.Sprintf("done updating %s stack", swarmStack.name))
+	stackStatus[swarmStack.name].Revision = stackMetadata.repoRevision
```
I did not change the `Revision` so that the current UI does not break.

sanzoghenzo reviewed (approved):
This looks great!

Only one note on the `updateStack` function: the logger is initialized to show the stack name as extra parameter, so there's no need to add it to the message... but it doesn't hurt, either :wink:

clangenb commented:
> This looks great!
> 
> Only one note on the `updateStack` function: the logger is initialized to show the stack name as extra parameter, so there's no need to add it to the message... but it doesn't hurt, either :wink:

Thanks a lot for the continuously reat feedback! Yeah, I noticed that, but I find it hard to see to which stack the logs belong to, as the logger appends the stack name instead of prepending. This is why I did it like that.

If you have a better idea, I am happy to listen. 😄

sanzoghenzo commented:
> I find it hard to see to which stack the logs belong to, as the logger appends the stack name instead of prepending. This is why I did it like that.
> 
> If you have a better idea, I am happy to listen. 😄

You're right, it is more readable, and I don't know go that well to propose better ways.

Now that all is good, I'll try to give it a spin before merging

sanzoghenzo commented:
I'm sorry I didn't have the time to test this, and I don't believe I can spend time on this project for a while...
@m-adawi should I go ahead and merge it anyway, or would you like to give it a spin first?

sanzoghenzo reviewed (changes requested):
Hi there, sorry for the radio silence!

I finally got some time to test it, but I realized there still are some things I would do differently, mainly to concentrate the configuration in a single place and to avoid high coupling in the code.

sanzoghenzo commented:
> Hey there, thanks for looking into this. I have put in some considerable effort into this, which goes beyond this PR.

Oh wow, now I see the other branches on your fork.. you've done a tremendous job!
Again, sorry for ignoring this 😅 

> I guess, it is better to extract some new PRs based on my current state if there is still interest in this. :)

Of course! Should we close this and start with a new PR? Or could you push your latest edits on this branch? (We will sqash everything in a single feature commit, so the lengthy commit history is not a problem!)

vhaudiquet commented:
This should fix #126, and I'm really grateful for your work ! :)

vhaudiquet commented:
@clangenb Any updates on this? I would really love to try that! Is the version on your branch ready or does it need more work? I could build it myself in the mean time

clangenb commented:
Thanks for remindinding me. I should have time this today. 😃

logoff commented:
hi. I was having the issue exposed in #126, and saw this PR. are there any news? this PR looks really promising :heart:

sanzoghenzo commented:
Hey there! as you may have noticed, I started taking some of the changes in your fork and made a pair of PRs with the simple features.

Now that I got to the matter of this PR, I was wondering: this completely disables the ability to reconcile drifting from the desired state declared in the git repository, for example if somebody removes a stack/service by hand, it is not re-deployed until an edit is made to the stack file, isn't it?

As I wrote before https://github.com/m-adawi/swarm-cd/pull/110#issuecomment-2707485460, we could solve this by adding a `--resolve-image changed` to the `docker stack deploy` command, and even better handling this with an option that can be set globally and overridden per-stack.

clangenb commented:
Good morning,

Amazing, thanks a lot for taking the work!

Yes, you are right in this regard. 

* If a stack is removed from the stacks file, the stack will be removed, and no longer deployed even if there are updates in the docker compose file. (I guess it should re-pull the image if the stack is re-introduced into the stacks file, though - not sure now from the top of my head.)
* If a stack is in the stacks file, the hash of the marshalled docker config file must have changed in order to trigger a redeployment. (Hence, comments in the docker compose file do not change the hash, as they are ignored during marshalling).

In general, I am not a big fan of implicitly updating docker images if the image behind the same tag changes, but I can see that short-term this can be wished for.

I think a global, and per stack config might make sense here. As the config is re-read upon every loop, we could simply add it for the time being when a stack should have its images updated.

sanzoghenzo commented:
> Yes, you are right in this regard.
> 
>     * If a stack is removed from the stacks file, the stack will be removed, and no longer deployed even if there are updates in the docker compose file. (I guess it should re-pull the image if the stack is re-introduced into the stacks file, though - not sure now from the top of my head.)

What I was trying to picture is a little bit different:

1. you deploy the stack with swarm-cd
2. you or another user runs `docker stack rm stack-name` manually
3. your stack is gone until you change something of one of the services in the stack

But your use case is also something to take into account.

clangenb commented:
Oh, I see what you mean. However, do not think that automatically adding it back should be done if the docker image behind a tag has been changed. I would consider this a very rare edge case where manual intervention is a good approach.

sanzoghenzo commented:
> Oh, I see what you mean. However, do not think that automatically adding it back should be done if the docker image behind a tag has been changed. I would consider this a very rare edge case where manual intervention is a good approach.

My point is that it should automatically go back up without any edits, just like ArgoCD and Flux do for kubernetes: the git repository is the source of truth, and the tool is here to ensure that there are no drifts from the desired state.

This PR completely disables this behavior, which is not what I would expect from a GitOps solution inspired by ArgoCD.

clangenb commented:
I see. I am not the DevOps Pro here, but I can understand that the GitOps config should be the single source of truth now that I think about this. Even more so, if this is the behaviour of other tools in the ecosystem. Thanks for teaching me that. :)
